afdd/utils.py

```diff
@@ -148,53 +155,85 @@ def load_timeseries(conn: Connection, graphInfoDF: pd.DataFrame, start_time: str
   with conn.cursor() as cur:
     cur.execute(query, timeseries_ids + [start_time, end_time])
     rows = cur.fetchall()
-    dict = {}
-    for id in timeseries_ids:
-      data = [row[0] for row in rows if row[1] == id]
-      dict[id] = data
     
-    result = pd.DataFrame(dict)
-    conn.commit()
-    logger.info(result)
+    # make a dataframe out of the query results
+    df = pd.DataFrame(rows, columns=["ts", "value", "timeseriesid"])
+    logger.info(df)
+
+    # convert the ts column to datetimes
+    df['ts'] = pd.to_datetime(df['ts'])
 
-  return result
+    # pivot the df to have ts as the index, timeseriesid as the columns and value as the values
+    df_pivoted = df.pivot(index='ts', columns='timeseriesid', values='value')
+    df_pivoted = df_pivoted.sort_index()
 
-metric_map = {
-  "average": pd.Series.mean,
-  "min": pd.Series.min,
-  "max": pd.Series.max
-}
+    conn.commit()
+    logger.info(df_pivoted)
+
+  return df_pivoted
+
+# normal comparison helpers
+# metric_map = {
+#   "average": pd.Series.mean,
+#   "min": pd.Series.min,
+#   "max": pd.Series.max
+# }
+
+# series comparison helpers
+def series_in_range(data, threshold: tuple):
+  return data.between(threshold[0], threshold[1])
+
+series_symbol_map = {
+    '>': pd.Series.gt,
+    '>=': pd.Series.ge,
+    '<': pd.Series.lt, 
+    '<=': pd.Series.le,
+    'in': series_in_range
+    }
+
+def series_comparator(op, data, threshold):
+  return series_symbol_map[op](data, threshold)
 
 # looping through point readings and checking for anomaly
 # only works for true or false rules
-def analyze_data(timeseries_data: pd.DataFrame, rules: List[Rule], start_time: str, end_time: str) -> List[tuple]:
+def analyze_data(timeseries_data: pd.DataFrame, rule: Rule) -> List[tuple]:
+  """
+  Evaluates the given timeseries data against the given rule and returns a list of tuples representing anomalies.
+  """
   anomaly_list = []
-  for ts_id, values in timeseries_data.items():
-    for rule in rules:
-      func = metric_map[rule.condition.metric]
-      sample_data = func(values)
-      threshold = rule.condition.threshold
-      logger.info(f"{ts_id}:{sample_data}, {threshold}")
-      op = rule.condition.operator
-      if comparator(op, sample_data, threshold):
-        anomaly = Anomaly(start_time=start_time, end_time=end_time, rule_id=rule.rule_id, value=sample_data, timeseriesid=ts_id)
+  if rule.condition.metric == "average":
```

Comment:
Should use Metric.average instead of string compare. Not sure if your starting this on a new branch